### PR TITLE
Improve CI some more

### PR DESCRIPTION
**Pull request by benfrankel**

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -11,86 +11,104 @@ env:
 
 jobs:
   # Run unit tests.
-  test:
-    name: Test Suite
+  unit-test:
+    name: Unit Tests
```
Comment (janhohenheim) [inline .github/workflows/ci.yaml:15]: Misnomer: also includes integration tests, regression tests, smoke tests, benchmarks, etc.
This comment is for all instances of this rename

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (benfrankel) [reply .github/workflows/ci.yaml:15]: Alright. What should I call this to distinguish it from doc tests?

Comment (janhohenheim) [reply .github/workflows/ci.yaml:15]: But it also runs doc tests! The second one is more of a `Check Docs` or `Validate Docs`

Comment (benfrankel) [reply .github/workflows/ci.yaml:15]: Okay. The manual page was confusing:
```
       --all-targets
           Test all targets. This is equivalent to specifying --lib --bins --tests --benches --examples.

       --doc
           Test only the library’s documentation. This cannot be mixed with other target options.
```

Comment (janhohenheim) [reply .github/workflows/ci.yaml:15]: Wait, hold up. You're right, it does not run doc tests

Comment (janhohenheim) [reply .github/workflows/ci.yaml:15]: `Tests` vs `Doc Tests` is good for me then.

Comment (benfrankel) [reply .github/workflows/ci.yaml]: https://github.com/rust-lang/cargo/issues/6669 :smiling_face_with_tear:

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -10,87 +10,87 @@ env:
...
-      - name: Run cargo test
-        run: cargo test --all-targets
+      - name: Populate target directory from cache
+        uses: Leafwing-Studios/cargo-cache@v1
+
+      - name: Run tests
+        run: |
+          cargo test --workspace --all-targets --all-features
+          # TODO: Workaround for https://github.com/rust-lang/cargo/issues/6669
+          cargo test --workspace --docs --all-features
```
Comment (janhohenheim) [inline .github/workflows/ci.yaml:35]:
```suggestion
          # Workaround for https://github.com/rust-lang/cargo/issues/6669
          cargo test --workspace --docs
```
This reads as if the user needs to do something

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (benfrankel) [reply .github/workflows/ci.yaml:35]: Yeah fair enough. This is how I mark it in my own projects so I know to remove the workaround when it's no longer necessary.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -10,87 +10,87 @@ env:
...
+      - name: Install dependencies
         run: sudo apt-get update; sudo apt-get install --no-install-recommends libasound2-dev libudev-dev
 
-      - name: Run cargo doc
-        run: cargo doc --no-deps
+      - name: Populate target directory from cache
+        uses: Leafwing-Studios/cargo-cache@v1
+
+      - name: Check documentation
+        run: cargo doc --workspace --all-features
```
Comment (janhohenheim) [inline .github/workflows/ci.yaml:96]:
```suggestion
        run: cargo doc --workspace --all-features --document-private-items
```
This is what we are interested in, after all

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -10,87 +10,87 @@ env:
...
-      - name: Run cargo test
-        run: cargo test --all-targets
+      - name: Populate target directory from cache
+        uses: Leafwing-Studios/cargo-cache@v1
+
+      - name: Run tests
+        run: |
+          cargo test --workspace --all-targets --all-features
+          # Workaround for https://github.com/rust-lang/cargo/issues/6669
+          cargo test --workspace --docs
```
Comment (janhohenheim) [inline .github/workflows/ci.yaml:35]:
```suggestion
          cargo test --workspace --doc
```

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (benfrankel) [reply .github/workflows/ci.yaml:35]: Oops.

**Review by janhohenheim — APPROVED** [review]

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -10,87 +10,87 @@ env:
...
+      - name: Install dependencies
         run: sudo apt-get update; sudo apt-get install --no-install-recommends libasound2-dev libudev-dev
 
-      - name: Run cargo doc
-        run: cargo doc --no-deps
+      - name: Populate target directory from cache
+        uses: Leafwing-Studios/cargo-cache@v1
+
+      - name: Check documentation
+        run: cargo doc --workspace --all-features --document-private-items
```
Comment (janhohenheim) [inline .github/workflows/ci.yaml:96]:
```suggestion
        run: cargo doc --workspace --all-features --document-private-items --no-deps
```

**Review by janhohenheim — CHANGES_REQUESTED** [review]

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Add `RUSTDOCFLAGS='-D warnings'` to the docs step.
You're about to whether there is a `--deny-warnings` or similar.
No: https://github.com/rust-lang/cargo/issues/8424

**Review by janhohenheim — APPROVED** [review]

**Pull request merged**